### PR TITLE
New histogram metric for snapshot duration

### PR DESCRIPTION
**Pull request by ArthurSens**
Signed-off-by: arthursens <arthursens2005@gmail.com>

**What this PR does / why we need it**:
This PR adds a new histogram `kubevirt_snapshot_duration_seconds` that provides detailed information about time spent on VMI snapshots.

This new metric can be used in the following use cases:
* Detect if snapshots are failing:
```
delta(kubevirt_snapshot_duration_seconds_count{status=”failed”}[5m]) > 1
```
* Detect average time spent on snapshots per status:
```
kubevirt_snapshot_duration_seconds_sum / kubevirt_snapshot_duration_seconds_count
rate(kubevirt_snapshot_duration_seconds_sum[5m]) / rate(kubevirt_snapshot_duration_seconds_count[5m]) #average time spent in last 5 minutes
```
* Detect average time spent on snapshot, ignoring status:
```
sum without (status) (kubevirt_snapshot_duration_seconds_sum) / 
sum without (status) (kubevirt_snapshot_duration_seconds_count)
```
* Calculate the 90th percentile in the last 5 minutes taking source node, target node, and status in consideration:
```
histogram_quantile(0.9, 
  rate(kubevirt_snapshot_duration_seconds_bucket[5m])
)
```
* Calculate the 90th percentile in the last 5 minutes ignoring status:
```
histogram_quantile(0.9, 
  sum without (status)(
    rate(kubevirt_snapshot_duration_seconds_bucket[5m])
  )
)
```

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
References used to create this PR:
* [How does a prometheus histogram work](https://www.robustperception.io/how-does-a-prometheus-histogram-work)
* [Prometheus client_golang examples](https://github.com/prometheus/client_golang/blob/master/examples/random/main.go)

**Release note**:

```release-note
Added new histogram "kubevirt_snapshot_duration_seconds" to the virt-controller
```

### CODE REVIEW

**Comment by ArthurSens** [comment]
Similar to #3887, still needs e2e tests and we need to come to a consensus regarding bucket's values

File: pkg/virt-controller/watch/snapshot.go
Comment (ashleyschuett) [inline pkg/virt-controller/watch/snapshot.go:797]: this function gets pretty nested. could we instead do 

```
if vmSnapshotPregressing(ss) {
   return 
}

....
```

File: pkg/monitoring/snapshots/prometheus/prometheus.go
Comment (dhiller) [inline pkg/monitoring/snapshots/prometheus/prometheus.go:44]: If we really want to do this we should create an issue for this and remove this comment.

**Review by dhiller — APPROVED** [review]
One nit found, also +1 for Ashleys comment. But you need to rebase anyway.

/approve

**Comment by ArthurSens** [comment]
/hold
/lgtm cancel
/approve cancel

This PR was created at a time when #3887 did not introduce any API changes for configuring histogram metrics.
I will address what ashley commented but I'd like to wait until #3887 is reviewed and merged since this PR will extend the API in a similar way

**Comment by fabiand** [comment]
@sradco @yuvalturg @rmohr what should we do about this PR - drop or keep and rebase?

**Comment by fabiand** [comment]
/unapprove 
?

**Pull request closed without merging**
